src/librustc_mir/interpret/eval_context.rs

```diff
@@ -16,28 +16,34 @@ use syntax::codemap::{self, DUMMY_SP, Span};
 use syntax::ast::{self, Mutability};
 use syntax::abi::Abi;
 
-use error::{EvalError, EvalResult};
-use lvalue::{Global, GlobalId, Lvalue, LvalueExtra};
-use memory::{Memory, MemoryPointer, TlsKey, HasMemory};
-use memory::Kind as MemoryKind;
-use operator;
-use value::{PrimVal, PrimValKind, Value, Pointer};
-use validation::ValidationQuery;
+use super::{
+    EvalError, EvalResult,
+    Global, GlobalId, Lvalue, LvalueExtra,
+    Memory, MemoryPointer, HasMemory,
+    Kind as MemoryKind,
+    operator,
+    PrimVal, PrimValKind, Value, Pointer,
+    ValidationQuery,
+    Machine,
+};
+
+pub struct EvalContext<'a, 'tcx: 'a, M: Machine<'tcx>> {
+    /// Stores data required by the `Machine`
+    pub machine_data: M::Data,
 
-pub struct EvalContext<'a, 'tcx: 'a> {
     /// The results of the type checker, from rustc.
-    pub(crate) tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    pub tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
     /// The virtual memory system.
-    pub(crate) memory: Memory<'a, 'tcx>,
+    pub memory: Memory<'a, 'tcx, M>,
 
     #[allow(dead_code)]
     // FIXME(@RalfJung): validation branch
     /// Lvalues that were suspended by the validation subsystem, and will be recovered later
     pub(crate) suspended: HashMap<DynamicLifetime, Vec<ValidationQuery<'tcx>>>,
```

Comment:
So the plan is to keep the validation stuff for CTFE, or to factor it out later?